### PR TITLE
feat(700): support join trigger

### PR DESCRIPTION
**Pull request by d2lam**
Support Join Trigger

Example:
```
A: 
   requires: [~B, C,D]
```

joinList of A = [C,D]
- When B is successful, trigger A (since B is not in the joinList)
- When C is successful, check if D is successful. If yes, trigger A. If no, do nothing. 

Related: https://github.com/screwdriver-cd/screwdriver/issues/700

### CODE REVIEW

File: plugins/builds/index.js [hunk 1]
```diff
@@ -70,6 +121,52 @@ exports.register = (server, options, next) => {
             });
     });
 
+    server.expose('triggerNextJobs', (config) => {
```
Comment (tkyi) [inline plugins/builds/index.js]: JSdoc?

File: plugins/builds/index.js [hunk 2]
```diff
@@ -70,6 +121,52 @@ exports.register = (server, options, next) => {
...
+        const eventFactory = server.root.app.eventFactory;
+        const jobFactory = server.root.app.jobFactory;
+        const buildFactory = server.root.app.buildFactory;
+        const { pipeline, job, build, username, scmContext } = config;
+        const currentJobName = job.name;
+        const pipelineId = pipeline.id;
+
+        return eventFactory.get({ id: build.eventId }).then((event) => {
+            const workflowGraph = event.workflowGraph;
+            const nextJobs = workflowParser.getNextJobs(workflowGraph, { trigger: job.name });
```
Comment (tkyi) [inline plugins/builds/index.js]: Can use `trigger: currentJobName`

File: plugins/builds/index.js [hunk 3]
```diff
@@ -7,6 +7,57 @@ const stepGetRoute = require('./steps/get');
...
+
+    return jobFactory.get({
+        name: jobName,
+        pipelineId
+    }).then((job) => {
+        if (job.state === 'ENABLED') {
+            return buildFactory.create({
+                jobId: job.id,
+                sha: build.sha,
+                parentBuildId: build.id,
```
Comment (tkyi) [inline plugins/builds/index.js]: Is `parentBuildId` needed?

Comment (tkyi) [reply plugins/builds/index.js]: Nvm I see it in the database.
We should update the docs here at some point: https://github.com/screwdriver-cd/models/blob/master/lib/buildFactory.js#L145-L156

Comment (d2lam) [reply plugins/builds/index.js]: Yep. parentBuildId is the build that triggers this one.

File: test/plugins/builds.test.js [hunk 4]
```diff
@@ -746,6 +747,161 @@ describe('build plugin test', () => {
                     });
                 });
             });
+
+            describe('join', () => {
```
Comment (jithin1987) [inline test/plugins/builds.test.js]: why is this test suite nested inside `buildToken` ?

Comment (minz1027) [reply test/plugins/builds.test.js]: The `PUT /builds/id` has two scopes, `user token` and `build token`. User can only update build status to `ABORTED`. Build can do all the stuff.

File: plugins/builds/index.js [hunk 5]
```diff
@@ -7,6 +7,57 @@ const stepGetRoute = require('./steps/get');
...
+        return null;
+    });
+}
+
+/**
+ * Check if all the jobs in joinList are successful
+ * @method isJoinDone
+ * @param  {Array}      joinList       array of jobs(name,id) that are in join
+ * @param  {Array}      finishedBuilds array of finished builds belong to this event
+ * @param  {Object}     workflowGraph  the event workflowGraph
```
Comment (minz1027) [inline plugins/builds/index.js]: Doesn't look like you need this

Comment (d2lam) [reply plugins/builds/index.js]: 👍  Was leftover from before!

**Pull request merged**